src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -84,10 +105,19 @@ internal class PackageRestoreDataSource : ChainedProjectValueDataSourceBase<Rest
             _fileSystem = fileSystem;
             _logger = logger;
             _projectDependentFileChangeNotificationService = projectDependentFileChangeNotificationService;
+            _solutionRestoreService4 = solutionRestoreService4;
         }
 
         protected override IDisposable? LinkExternalInput(ITargetBlock<IProjectVersionedValue<RestoreData>> targetBlock)
         {
+            // Register before this project receives any data flows containing possible nominations.
+            // This is needed because we need to register before any nuget restore or before the solution load.
+#pragma warning disable RS0030 // Do not used banned APIs
+            var registerRestoreInfoSourceTask = Task.Run(() => _solutionRestoreService4.RegisterRestoreInfoSourceAsync(this, _projectAsynchronousTasksService.UnloadCancellationToken));
```

Comment:
There is no guarantee this Task.Run completes before we've done the first nomination, is that the intention?

Reply:
@davkean has a good point, the Task.Run might introduce more delay, and maybe an extra try/catch and report the exception when RegisterRestoreInfoSourceAsync throws is a better choice. Generally, I think it is a reasonable starting point, but my feeling is that the registration need be moved to another location later.  Here are a few design rules around this:
1, there is no need to control order between the registration of the service, and first nomination of the same project, because we don't block the nomination from the same project (which will lead a deadlock, because the nomination blocks dataflow to process new data). So the order to finish this task & chain data sources in this function has no logic meaning.

2, we should not wait RegisterRestoreInfoSourceAsync to finish.  We want to call it earlier, but there is no need to wait it to finish until the project is closed. The point is that we don't delay our logic due to some sync logic inside nuget implementation.  Their original draft contains logic to switch to UI thread -- basically to retrieve project info from the solution. It leads into problems, because we may call solution fully initialized, and it runs into some odd state. NuGet will block the process to wait solution to be loaded, so waiting the process will delay project loading logic.  My earlier suggestion to them is that Register function doesn't need to be async. Whatever nuget need to do, it can be done in their inner queue, which need be emptied before nuget restore to start. (The queue acts like a virtual RestoreInfoSource itself.) But i decided that pushing too much would further delay this work, so i think it is better to get something first.

3, we need call RegisterRestoreInfoSourceAsync (not wait to finish) before the first restore may start. A simple choice would be to register that before project is fully loaded (or before solution load finishes). The point is that NuGet should never start restore until the solution is fully loaded, otherwise it won't be able to compute dependencies. Also we need register restore data source to operation progress on that point as well. (I doubt it might be registered later, so the operation progress can finish earlier when DTBB cache is ready sometime, but it might be wrong. The reason is that I saw operation progress finished immediately after solution was loaded, and package restore message in the status bar later sometime in warm loading.)  Because of that, I doubt this is the right location to call RegisterRestoreInfoSourceAsync.  I looped @davkean in for this issue in an earlier comments in this PR.  (I don't want to block this PR for this, because it might not be a major problem in real world, and getting it to test might be more important at this point.)